### PR TITLE
Feature/type checked

### PR DESCRIPTION
**Pull request by SebastianJL**
Hi please review and test. Feel free to add comments or documentation, write new unittests or use in other projects to see whether it works properly. Also there is a TODO section which describes task that are still left to do. Add other task if neccessary.

## Dynamic Type Checking
This pull request introduces a decorators module with type_checked as its only decorator so far. Type_checked is a decorator for dynamic type checking based on annotations.

## Features
From the docstring:
 - Decorates functions, static methods and bound methods.
 - Each Parameter and the return value can be annotated with one expression of <class 'type'>.
 - Other expressions or missing annotations are ignored.
 - Default values are type checked once when the function is decorated.
 - All other annotations are checked once per function call.
 - Type checking of arguments, default values or the return value can be turned off separately.
 - Annotations for *args and **kwargs are ignored.
 - The effect of this decorator is turned of if \_\_debug\_\_ == False i.e. if python is run with the option -O.

## TODO
- [x] Only works for functions. Implement for bound methods.
- [x] Test for bound methods.
- [x] Test for static methods.
- [x] Test for class methods.
- [x] Add Travis CI DocTests

---
This change is [Reviewable](https://reviewable.io/reviews/phynix/uniphy/12)

### CODE REVIEW

**Comment by jonas-eschle** [comment]
:D sorry for the codacy-bot spam, still have to adjust some "hard-to-spot" settings.

two things came to my mind:
- may remove the readme from the commit and replace it with the current from the master (which is basically empty)
- we may should consider adding deeper structure to the repository and create a "tool" folder (or similar, or even deeper). I think this would greatly add to the maintainability and separation of things instead of keeping the files too flat.

**Comment by SebastianJL** [comment]
+ I guess we should finish the pull request for the README.md first?
+ Does the deeper folder structure make sense at this stage? I mean a refactor from decorators to tools.decorators is pretty easy.

**Comment by jonas-eschle** [comment]
* agree
 * true, although with other modules I think this becomes a necessity. Probably this also allows for small yet useful tools to be placed there (like to_list conversion that converts a single string to [string] instead of [s,t,r,i,n,g]) and such things. But I agree, this refactoring is an easy thing (with pycharm :D) and can be done later. It's just nice to have a future place in mind.

**Comment by SebastianJL** [comment]
@mayou36 
How did you resolve the problem with codacy reporting a syntax error?

**Comment by jonas-eschle** [comment]
Either it resolved itself or because I simply told it to ignore the "issue". I did the second, but not sure whether probably the first happend as well.

**Comment by SebastianJL** [comment]
@mayou36 
Can you rerun travis CI? Somehow it failed but apparently it didn't use the settings you used in your last commit. The uniphy/\*{.py,/\*{.py,/\*.py}} part is somehow still in although you changed it to /*.py.

**Comment by SebastianJL** [comment]
@mayou36 Ok. Did it by closing and reopening the pull request. I think we are only missing the documentation at this point. Do you think you can manage to auto-generate it?

**Comment by jonas-eschle** [comment]
@SebastianJL ok nice! I could also manually start it again, or an "empty" commit would work as well. The API autogeneration should work, I would prefer to set it up in order to work for anything in the repo, which may requires a small pipeline to have. But I will do that.

To me the code looks good, did you reconsider anything regarding the code snippet with the decorated decorators?

Btw I changed something in the travis build: Instead of using the default python env of Ubuntu (which is the travis build image) I changed it to install a conda environment and run it inside. Works so far (used it in other repos as well).

**Comment by SebastianJL** [comment]
Ok I had a look at the "decorator decorator" but it turns out that if anything like this should be done it needs more research (e.g. there would also be the possibility of  a decorator base class). At this point however this would go too far in my opinion. Therefore I'd say if you can review the pull request I will be happy with this.

**Comment by jonas-eschle** [comment]
I agree, sure will look through it

File: tests/test_decorators.py [hunk 1]
```diff
@@ -25,14 +28,14 @@ def test_func2(a, b: str, c=None, d: list = None):
...
     (1, 1, [], []),  # first annotation wrong
     (1, "hello", [], 1),  # second annotation wrong
-    (1, 1, 1, 1)  # both annoations wrong
+    (1, 1, 1, 1)  # both annotations wrong
 ]
 
-correct2 = {'a': 1, 'b': "hello", 'c': [], 'd': []}
+correct2 = OrderedDict(a=1, b="hello", c=[], d=[])
 failures2 = [
-    {'a': 1, 'b': 1, 'c': [], 'd': []},  # first annotation wrong
```
Comment (jonas-eschle) [inline tests/test_decorators.py]: Changed to OrderedDicts, as we rely on the ordering for the args. The ordering, although currently given in CPython, is not a feature but unreliable implementation detail, so OrderedDicts are the way to go

File: uniphy/decorators.py [hunk 2]
```diff
@@ -71,12 +72,12 @@ def __new__(cls, *args, **kwargs):
         elif args and callable(args[0]):
             # Called as @type_checked
             self.__init__()
-            return self.__call__(args[0])
```
Comment (jonas-eschle) [inline uniphy/decorators.py]: Why not just call self directly? ;)

**Review by jonas-eschle — COMMENTED** [review]
Looks good to me! Only found some typos/small fixes. I have tried to run the automatic checker of pycharm, it had of course a couple of false positives. Instead of completely ignoring the test-files, I added supressions (which works quite easy). This, on the other side, adds some additional comments, I don't think they are a problem, mostly not in the tests, but they can also be removed. Thought we may give it a try.

File: uniphy/decorators.py (hunk 2 above)
Comment (SebastianJL) [reply uniphy/decorators.py]: I cracked a smile when I saw that :)

File: tests/test_decorators.py (hunk 1 above)
Comment (SebastianJL) [reply tests/test_decorators.py]: :( I always forget about that.

**Comment by SebastianJL** [comment]
Some changes I made (this time it's final, I swear):
 - The global variables and functions in the decorators test file are now attributes to the test class.
 - Unittest Subtests have been used where sensible.
 - AssertRaises has been replaced with AssertRaisesRegex where sensible.

**Comment by SebastianJL** [comment]
@mayou36 Please approve pull request.

**Review by jonas-eschle — APPROVED** [review]
Looks good!

**Pull request merged**